graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -1637,37 +1664,99 @@ def _create_split_files_from_rates(
+            else:
+                # TODO: add support for sums <= 1.0, useful for large-scale link prediction
+                if sum(split_rates.tolist()) != 1.0:
+                    raise RuntimeError(f"Provided split rates  do not sum to 1: {split_rates}")
 
-        rng = default_rng(seed=seed)
+            split_list = split_rates.tolist()
+            logging.info(
+                "Creating split files for label column '%s' with split rates: %s",
+                label_column,
+                split_list,
+            )
 
-        # We use multinomial sampling to create a one-hot
-        # vector indicating train/test/val membership
-        def multinomial_sample(label_col: str) -> Sequence[int]:
-            if label_col in {"", "None", "NaN", None}:
-                return [0, 0, 0]
-            return rng.multinomial(1, split_list).tolist()
+            rng = default_rng(seed=seed)
+
+            # We use multinomial sampling to create a one-hot
+            # vector indicating train/test/val membership
+            def multinomial_sample(label_col: str) -> Sequence[int]:
+                if label_col in {"", "None", "NaN", None}:
+                    return [0, 0, 0]
+                return rng.multinomial(1, split_list).tolist()
+
+            group_col_name = "sample_boolean_mask"  # TODO: Ensure uniqueness of column?
+
+            # TODO: Use PandasUDF and check if it is faster than UDF
+            split_group = F.udf(multinomial_sample, ArrayType(IntegerType()))
+            # Convert label col to string and apply UDF
+            # to create one-hot vector indicating train/test/val membership
+            input_col = F.col(label_column).astype("string") if label_column else F.lit("dummy")
+            int_group_df = input_df.select(split_group(input_col).alias(group_col_name))
+            train_mask_df = int_group_df.select(F.col(group_col_name)[0].alias("train_mask"))
+            val_mask_df = int_group_df.select(F.col(group_col_name)[1].alias("val_mask"))
+            test_mask_df = int_group_df.select(F.col(group_col_name)[2].alias("test_mask"))
+        else:
+            # custom node/edge label
+            # create custom mask dataframe for one of the types: train, val, test
+            def process_custom_mask_df(input_df, split_file, mask_type):
+                if mask_type == "train":
+                    file_path = split_file.train
+                elif mask_type == "val":
+                    file_path = split_file.valid
+                elif mask_type == "test":
+                    file_path = split_file.test
+                else:
+                    raise ValueError("Unknown mask type")
+
+                if len(split_file.column) == 1:
+                    custom_mask_df = self.spark.read.parquet(
+                        self.input_prefix + "/" + file_path
+                    ).select(col(custom_split_file.column[0]).alias(f"custom_{mask_type}_mask"))
+                    mask_df = input_df.join(
```

Comment:
A join causes a shuffle. Are we certain the produced data order is not affected by this? There is a strong assumption that the order of the feature DataFrames will be same as for the structure

Reply:
I think we can not 100% make sure if they are in the same order. But I suppose the UDF will help to relieve the shuffling issue. I will change the logic to UDF.

Reply:
Let's keep it with the join for now and we can do some tests with mid size datasets to ensure the order of the output is correct.